source/common/tcp_proxy/tcp_proxy.h

```diff
@@ -462,6 +470,92 @@ class Filter : public Network::ReadFilter,
   };
 
   StreamInfo::StreamInfo& getStreamInfo();
+  class HttpStreamDecoderFilterCallbacks : public Http::StreamDecoderFilterCallbacks,
+                                           public ScopeTrackedObject {
+  public:
+    HttpStreamDecoderFilterCallbacks(Filter* parent);
+    // Http::StreamDecoderFilterCallbacks
+    OptRef<const Network::Connection> connection() override {
+      return parent_->read_callbacks_->connection();
+    }
+    StreamInfo::StreamInfo& streamInfo() override { return parent_->getStreamInfo(); }
+    const ScopeTrackedObject& scope() override { return *this; }
+    Event::Dispatcher& dispatcher() override {
+      return parent_->read_callbacks_->connection().dispatcher();
+    }
+    void resetStream(Http::StreamResetReason, absl::string_view) override {
+      PANIC("not implemented");
+    };
+    Router::RouteConstSharedPtr route() override { return nullptr; }
+    Upstream::ClusterInfoConstSharedPtr clusterInfo() override {
+      return parent_->cluster_manager_.getThreadLocalCluster(parent_->route_->clusterName())
+          ->info();
+    }
+    uint64_t streamId() const override {
+      return parent_->config_->tunnelingConfigHelper()->streamId();
+    }
+    Tracing::Span& activeSpan() override { return parent_->active_span_; }
+    OptRef<const Tracing::Config> tracingConfig() const override {
+      return makeOptRef<const Tracing::Config>(parent_->tracing_config_);
+    }
+    void continueDecoding() override {}
+    void addDecodedData(Buffer::Instance&, bool) override {}
+    void injectDecodedDataToFilterChain(Buffer::Instance&, bool) override {}
+    Http::RequestTrailerMap& addDecodedTrailers() override { PANIC("not implemented"); }
+    Http::MetadataMapVector& addDecodedMetadata() override {
+      static Http::MetadataMapVector metadata_map_vector;
+      return metadata_map_vector;
+    }
+    const Buffer::Instance* decodingBuffer() override { return nullptr; }
+    void modifyDecodingBuffer(std::function<void(Buffer::Instance&)>) override {}
+    void sendLocalReply(Http::Code, absl::string_view,
+                        std::function<void(Http::ResponseHeaderMap& headers)>,
+                        const absl::optional<Grpc::Status::GrpcStatus>,
+                        absl::string_view) override {}
+    void encode1xxHeaders(Http::ResponseHeaderMapPtr&&) override {}
+    Http::ResponseHeaderMapOptRef informationalHeaders() const override { return {}; }
+    void encodeHeaders(Http::ResponseHeaderMapPtr&&, bool, absl::string_view) override {}
+    Http::ResponseHeaderMapOptRef responseHeaders() const override { return {}; }
+    void encodeData(Buffer::Instance&, bool) override {}
+    void encodeTrailers(Http::ResponseTrailerMapPtr&&) override {}
+    Http::ResponseTrailerMapOptRef responseTrailers() const override { return {}; }
+    void encodeMetadata(Http::MetadataMapPtr&&) override {}
+    void onDecoderFilterAboveWriteBufferHighWatermark() override {}
```

Comment:
I think we need to hook up and test flow control e2e

I would be tentatively OK with a TODO For this also called out by the reloadable flag